Sergey Rutkovskiy

spectraise opened the pull request:
Python Final Task

- [x] First iteration

1. - [x]   -h, --help     show this help message and exit
2. - [x]   --version      Print version info
3. - [x]   --json         Print result as JSON in stdout
4. - [x]   --verbose      Outputs verbose status messages
5. - [x]   --limit LIMIT  Limit news topics if this parameter provided

- [x] Second iteration
- [x] Third iteration
- [x] Fourth iteration
- [x] Fifth iteration
- [ ] Sixth iteration

spectraise committed:
Basic functionality implemented

spectraise committed:
Added __init__.py file

spectraise committed:
Added setup.py file

spectraise committed:
Fixing imports

spectraise committed:
Fixed imports and moved the entry point to a separate function

spectraise committed:
Version 0.2

spectraise committed:
Added requirements.txt

spectraise committed:
Added description

spectraise committed:
Reworked project structure

1. Renamed 'src' package to 'rss_reader'
2. Modules 'feed', 'news' and 'parser' moved from package 'src' (new 'rss_reader') to package 'components' to avoid package and module name conflicts
3. Moved setup.py file to package 'src' (new 'rss_reader')

spectraise committed:
Added package 'lxml' to requirements

Shurke commented:
Исправьте пожалуйста конфликт, чтобы корректно заработал CI.

spectraise committed:
Updated requirements.txt to resolve conflict

spectraise committed:
Imports aligned to PEP8

spectraise committed:
Added __init__.py to 'components' package

spectraise committed:
Added tests

1. Added tests for arguments '--version', '--limit', '--json'
2. Function 'main' in 'rss_reader.py' now accepts the list of command line arguments
3. Method 'parse_args' in module 'parser' now accepts the list of command line arguments (accidentally added with the commit '09aca0cc5cfa39ef869663642fb546dc2050ca95' but not added to commit comment)

spectraise committed:
Added .travis.yml

spectraise committed:
Added README.md

Shurke commented:
Покрытие тестами слишком низкое.
nose.plugins.cover: ERROR: TOTAL Coverage did not reach minimum required: 50%

spectraise committed:
Updated .travis.yml

Now nosetests checks coverage only for packages 'rss_reader' and 'components'

spectraise committed:
Updated .travis.yml

Now nosetests checks coverage only for package 'rss_reader'

Shurke commented:
К сожалению, пока что не работает.
![Sergey Rutkovski](https://user-images.githubusercontent.com/44495358/117577471-8f702e00-b0f2-11eb-8728-b38472b3c654.png)

spectraise committed:
Reworked project structure

1. Module 'rss_reader' moved to package 'rss_reader'
2. Added package 'rss_reader' to setup.py

spectraise committed:
Fixed imports

spectraise committed:
Fixed imports

Shurke commented:
Если на вход rss_reader подать пустую стороку, то ничего не происходит.
На мой взгляд, пользователю стоит сообщить о том, что он забыл ввести адрес или что пустая строка не валидна. Либо любое другое сообщение на ваш вкус.
![Screenshot from 2021-05-10 11-06-26](https://user-images.githubusercontent.com/44495358/117626450-c93b4600-b17f-11eb-9ebe-c8bfbaf981d0.png)

Shurke commented:
Если ввести отрицательный лимит, то происходит что-то странное - выводится огромное количество записей.
Думаю, что так не должно быть. Также, стоит подумать о том, как обработать лимит раный 0.
![Screenshot from 2021-05-10 11-06-07](https://user-images.githubusercontent.com/44495358/117626790-2505cf00-b180-11eb-89fc-5c0c656d0318.png)

Shurke commented:
Стоит предусмотреть обработку ввода неверных адресов.
![Screenshot from 2021-05-10 11-12-34](https://user-images.githubusercontent.com/44495358/117627252-a52c3480-b180-11eb-8c93-087685c5d0e8.png)

Shurke commented on rss_reader/components/feed.py:
```diff
@@ -0,0 +1,43 @@
+"""This module contains a class that represent a feed"""
+
+import json
+from .news import News
```
По РЕР8 рекомендуется абсолютное импортирование, так как оно обычно более читаемо и ведет себя лучше (или, по крайней мере, даёт понятные сообщения об ошибках) если импортируемая система настроена неправильно.

spectraise replied:
Учту. Будет исправлено.

Shurke commented on rss_reader/components/feed.py:
```diff
@@ -0,0 +1,43 @@
+        This method creates news objects and adds them to the news list
+        :return: None
+        """
+        self.logger.info(' Preparing a feed')
```
Затесался лишний пробельчик.
Как и в других местах логирования.

spectraise replied:
![image](https://user-images.githubusercontent.com/82844906/117634933-14595700-b188-11eb-8eb7-b9b341dbe9c2.png)
Он не лишний, мне показалось что с пробелом более читаемо.

Shurke replied:
Тогда резонно. Вопросов больше нет.

Shurke commented on rss_reader/components/feed.py:
```diff
@@ -0,0 +1,43 @@
+    def __create_feed(self):
+        """
+        This method creates news objects and adds them to the news list
+        :return: None
```
Если метод/функция ничего не возвращает, то об этом можно не писать.

spectraise replied:
Учту.

Shurke commented on rss_reader/components/feed.py:
```diff
@@ -0,0 +1,43 @@
+    def __str__(self):
+        """
+        This method override default __str__ method which computes the string representation of an object
+        :return: str
```
Если вы хотите обозначить, какой тип возвращается, то наверное все же стоит воспользоваться аннотацией типов.
Либо стоит написать, чем является выходная строка.

spectraise replied:
Учту.

Shurke commented on rss_reader/rss_reader/rss_reader.py:
```diff
@@ -0,0 +1,46 @@
+"""This module is the entry point to the utility"""
+
+import logging
+import sys
+import requests
+from bs4 import BeautifulSoup
+from components.parser import Parser
+from components.feed import Feed
```
Импорты должны быть сгруппированы в следующем порядке:

1. импорты из стандартной библиотеки
2. импорты сторонних библиотек
3. импорты модулей текущего проекта

Вставляйте пустую строку между каждой группой импортов.

spectraise replied:
С группировкой вроде все верно. Пустые строки добавлю.

Shurke commented on rss_reader/rss_reader/rss_reader.py:
```diff
@@ -0,0 +1,46 @@
+        limit = args.limit
+        if limit:
+            limit = int(limit)
+        try:
+            logger.info(' Sending GET request to the specified URL')
+            response = requests.get(source_url)
+        except requests.exceptions.ConnectionError:
+            logger.error(' An error occurred while sending a GET request to the specified URL. Check the specified URL'
+                         ' and your internet connection')
+        else:
+            logger.info(' Parsing XML from the specified URL')
+            soup = BeautifulSoup(response.content, 'lxml-xml')
+            feed_title = soup.find('title').text
```
Как выяснилось, данная часть не может обработать случай с неверным адресом.
Кроме того, за счет наличия кодов статуса, можно сделать более расширенную обработку ответов от новостных серверов, предоставляя пользователю больший спектр информационных сообщений.

Shurke commented on rss_reader/tests/test_arguments.py:
```diff
@@ -0,0 +1,67 @@
+import unittest
+from unittest.mock import patch
+from io import StringIO
+import re
+import json
+from rss_reader.rss_reader.rss_reader import main
```
Импорты должны быть сгруппированы в следующем порядке:

1. импорты из стандартной библиотеки
2. импорты сторонних библиотек
3. импорты модулей текущего проекта

Вставляйте пустую строку между каждой группой импортов.

Shurke commented on rss_reader/tests/test_arguments.py:
```diff
@@ -0,0 +1,67 @@
+
+class TestRssReader(unittest.TestCase):
+    @patch('sys.stdout', new_callable=StringIO)
+    def test_version(self, mock_stdout):
```
Было бы неплохо добавить к каждому тесту строку документации с описанием того, что этот тест все таки делает то.

spectraise replied:
Учту. Будут добавлены.

Shurke commented on rss_reader/rss_reader/rss_reader.py:
```diff
@@ -0,0 +1,46 @@
+    logger = logging
+    if args.source:
+        source_url = args.source
+        limit = args.limit
+        if limit:
+            limit = int(limit)
```
В чем смысл данного участка кода?

spectraise replied:
В нем нет смысла. Изначально для аргумента --limit не был указан type=int, а после его добавления я забыл убрать этот фрагмент. Будет убран.

Shurke commented on rss_reader/rss_reader/rss_reader.py:
```diff
@@ -0,0 +1,46 @@
+    """This function is a entry point"""
+    parser = Parser()
+    args = parser.parse_args(argv)
+    visibility = args.verbose
+    to_json = args.json
```
Для чего вы создаете дополнительные переменные, если можно на прямую пользоваться значениями из `args`?

spectraise replied:
Учту. Будут убраны.

Shurke commented on rss_reader/rss_reader/rss_reader.py:
```diff
@@ -0,0 +1,46 @@
+        logging.basicConfig(level=logging.INFO)
+    else:
+        logging.basicConfig(level=logging.ERROR)
+    logger = logging
```
Это все можно в одну строку уместить.
```python
logger = logging.basicConfig(level=logging.INFO) if args.verbose else logging.basicConfig(level=logging.ERROR)
```

spectraise replied:
![image](https://user-images.githubusercontent.com/82844906/117637679-b712d500-b18a-11eb-8f35-8d2a6e4b0af7.png)
Конкретно как вы написали - нельзя, но можно уменьшить до двух строк. Но все же мне кажется что мой вариант более читаемый.

Shurke replied:
Справедливо.

Shurke commented on rss_reader/rss_reader/rss_reader.py:
```diff
@@ -0,0 +1,46 @@
+        logging.basicConfig(level=logging.ERROR)
+    logger = logging
+    if args.source:
+        source_url = args.source
+        limit = args.limit
```
Опять же, для чего вводить дополнительные переменные?

spectraise replied:
Учту. Будут убраны.

spectraise committed:
Changed rss_reader.py

1. Removed unnecessary variables. Now the values are taken directly from 'args'.
2. Imports are grouped and separated with blank lines.
3. Added errors for empty source URL and --limit less than one

spectraise committed:
Merge remote-tracking branch 'origin/main'

spectraise committed:
Imports are grouped and separated with blank lines.

spectraise committed:
Changed documentation

spectraise committed:
Added error for the case when the specified URL does not contain RSS

spectraise committed:
Added error for the case when the specified URL has no schema supplied

spectraise committed:
Added documentation to tests

spectraise committed:
Added tests for errors

spectraise committed:
Imports are grouped and separated with blank lines.

spectraise committed:
[Version 0.3] Added news caching

1. Added method '__create_feed_from_cache' to module 'feed'
2. Method '__parse_date' in module 'news' now returns datetime object (previously it returned a formatted date string)
3. Added method '__cache_news' to module 'news'
4. Added method '__from_cache' to module 'news'
5. Added command-line argument '--date'
6. JSON structure changed

spectraise committed:
Functionality for caching has been moved to module 'cache'

1. Removed method '__create_feed_from_cache' from module 'feed' (the functionality of this method is now implemented in the 'cache' module)
2. Removed method '__cache_news' from module 'news' (the functionality of this method is now implemented in the 'cache' module)

spectraise committed:
Updated 'Usage'

spectraise committed:
Updated documentation for the 'cache' module

spectraise committed:
[Version 0.4] Added conversion to PDF and HTML format

1. JSON structure changed
2. Added command-line arguments '--to-pdf' and '--to-html'
3. Added validator for command-line argument '--date'
4. Added method '__parse_enclosure' to module 'news'
5. Removed unused imports from module 'feed'
6. Added method '__cache_images' to module 'cache'
7. Added module 'converter'

spectraise committed:
Fixed test for argument --version

spectraise committed:
Fixed W605 invalid escape sequence '\d'

spectraise committed:
Fixed an error that occurs when passing the '/' value to argument '--to-pdf' or '--to-html'

Shurke commented:
Как замечание, стоит добавить более понятное для пользователя сообщение о том, что новостей на запрашиваемую дату нет.
![Screenshot from 2021-05-21 18-13-31](https://user-images.githubusercontent.com/44495358/119159876-6049ad00-ba60-11eb-9d02-c9282ebcd1e6.png)

Shurke commented:
Конвертация в файл не рабоатет, когда хотим взять новости из архива или сразу из ленты.
> `--date `should also work correctly with format converter and to not require internet access.
![Screenshot from 2021-05-21 18-21-22](https://user-images.githubusercontent.com/44495358/119160915-68561c80-ba61-11eb-83c1-77a796590092.png)

![Screenshot from 2021-05-21 18-18-57](https://user-images.githubusercontent.com/44495358/119160526-0d242a00-ba61-11eb-976d-a1fd13d37c72.png)

Shurke commented on rss_reader/components/cache.py:
```diff
@@ -0,0 +1,144 @@
+import os
```
> Docstrings are mandatory for all methods, classes, functions and **modules**.

Shurke commented on rss_reader/components/cache.py:
```diff
@@ -0,0 +1,144 @@
+                                                     'items': {
+                                                         0: news.to_dict()
+                                                     }}
+                self.__write_cache(cache_file_path, cached_data)
+                self.__cache_images(news)
```
Данные строки повторяются (56-57 и 64-65).
Может стоит подумать над тем, чтобы избежать дублирования?

spectraise commented:
> Конвертация в файл не рабоатет, когда хотим взять новости из архива или сразу из ленты.
> > `--date `should also work correctly with format converter and to not require internet access.
> ![Screenshot from 2021-05-21 18-21-22](https://user-images.githubusercontent.com/44495358/119160915-68561c80-ba61-11eb-83c1-77a796590092.png)
> 
> ![Screenshot from 2021-05-21 18-18-57](https://user-images.githubusercontent.com/44495358/119160526-0d242a00-ba61-11eb-976d-a1fd13d37c72.png)
> 

В --to-html и --to-pdf должен передаваться путь по которому он должен быть сохранен. У вас ничего не передается.

Shurke reviewed (commented):
> В --to-html и --to-pdf должен передаваться путь по которому он должен быть сохранен. У вас ничего не передается.

Тогда стоит сообщить пользователю о том, что он забыл путь к файлу.

Shurke commented on rss_reader/components/cache.py:
```diff
@@ -0,0 +1,144 @@
+            os.mkdir(self.cache_folder_path)
+        self.cache_images_folder_path = self.cache_folder_path + 'images' + os.path.sep
+        if not os.path.exists(self.cache_images_folder_path):
+            os.mkdir(self.cache_images_folder_path)
```
Если я все правильно понял, то без `self.cache_folder_path` не будет и `self.cache_images_folder_path`.
Тогда во втором `if` нет смысла. и после создания `self.cache_folder_path` следует создавать `self.cache_images_folder_path`.

Shurke commented on rss_reader/components/cache.py:
```diff
@@ -0,0 +1,144 @@
+                    source_feed_already_cached = True
+                    source_feed_index = cached_feed_index
+                    for cached_news in cached_feed['items'].values():
+                        if cached_news['url'] == news.link:
```
Если я правильно понимаю, то вы используете URL новости, как уникальные идентификатор новости.
Если мое предположение верно, то вместо цикла можно через list comprehension (а лучше tuple comprehension) собрать все URLs из кеша и просто проверять, входит ли URL новости в кеш или нет.

Shurke commented on rss_reader/components/cache.py:
```diff
@@ -0,0 +1,144 @@
+        source_feed_index = None
+        source_feed_cached_news_number = None
+        cached_data = self.__get_cached_data(cache_file_path)
+        if cached_data:
+            for cached_feed_index, cached_feed in cached_data.items():
+                if cached_feed['source'] == news.source:
+                    source_feed_already_cached = True
+                    source_feed_index = cached_feed_index
+                    for cached_news in cached_feed['items'].values():
+                        if cached_news['url'] == news.link:
+                            news_already_cached = True
+                            break
+                    if not news_already_cached:
+                        source_feed_cached_news_number = len(cached_feed['items'])
+                if news_already_cached or source_feed_already_cached:
+                    break
+            if not news_already_cached:
+                if source_feed_already_cached:
+                    cached_data[source_feed_index]['items'][source_feed_cached_news_number] = news.to_dict()
+                else:
+                    cached_data[len(cached_data)] = {'title': news.feed_title,
+                                                     'source': news.source,
+                                                     'items': {
+                                                         0: news.to_dict()
```
А как себя поведет ридер, если мы в разные дни запросим одну и туже ленту?

spectraise replied:
Если честно, я не совсем понимаю ваш вопрос.

Shurke replied:
Предположим, я с помощью вашего ридера буду запрашивать новости 22, 24 и 26 мая.
Как будут сохраняться новости?

spectraise replied:
В этом случае ридер делает следующее:
Для каждой новости из фида вызывается Сache.cache_news -> Внутри этого метода определяется название файла с кэшем (дата публикации новости в формате "%Y%m%d" с расширением .json) -> Если такого файла не существует или он пустой, то ридер создает словарь фида с одной новостью. Если же такой файл существует, то ридер проверяет есть ли в нем кэш для фида из которого была получена новость -> Если кэша нет, то ридер добавляет в данные кэша еще один фид с одной новость. Если кэш для фида есть, то ридер определяет количество новостей в этом фиде и в конец фида добавляет новую новость (если она до этого не была кэширована ) -> Далее измененные данные кэша записываются в файл (если до этого новость не была кэширована) -> После этого ридер ищет в ссылках новости изображения и скачивает их (если до этого новость не была кэширована)

Shurke commented on rss_reader/components/cache.py:
```diff
@@ -0,0 +1,144 @@
+                    cached_feeds[len(cached_feeds)] = temp_feed
+                    temp_feed = None
+                if len(news_list) == limit:
+                    break
```
Стоит на этот участок кода посмотреть еще раз.
Я думаю, что тут есть варианты для улучшения.
Сейчас это практически не реально понять.

Shurke commented on rss_reader/components/converter.py:
```diff
@@ -0,0 +1,171 @@
+import glob
+import urllib.request
+import urllib.error
+import imghdr
```
Напомню, что импорты располагаются в алфавитном порядке.

Shurke commented on rss_reader/components/converter.py:
```diff
@@ -0,0 +1,171 @@
+            self.logger.info(f' PDF file created and saved at {result_path + result_filename}')
+        except PermissionError:
+            self.logger.error(
+                f' Unable to save PDF file at {result_path + result_filename}. Permission denied.')
```
Глядя на эту функцию есть несколько идей:
1. Можно подумать о том, чтобы сделать отдельные классы-конвертеры для pdf и html. Я думаю, что можно выделить общие методы конвертации в базовый класс `Converter` и добавить уже специализированные в `PdfConverter` и `HtmlConverter` (как вариант).
2. Пересмотреть `to_pdf` и `to_html`. Что-то мне подсказывает, что в этих методах есть общие части кода, которые можно вынести в отдельные методы и переиспользовать.

Shurke commented on rss_reader/components/converter.py:
```diff
@@ -0,0 +1,171 @@
+            self.logger.info(f' HTML file created and saved as {result_path + result_filename}')
+        except PermissionError:
+            self.logger.error(
+                f' Unable to save HTML file at {result_path + result_filename}. Permission denied.')
```
Сморите комментарий к `to_pdf`.

Shurke commented on rss_reader/components/feed.py:
```diff
@@ -0,0 +1,53 @@
+                                       }}}, indent=4, ensure_ascii=False)
+        else:
+            self.logger.info(' Printing news to STDOUT')
+            return '\n\n'.join(str(news) for news in self.news_list[:self.limit])
```
А что если попробовать
```python
if self.to_json and self.date:
    ...
elif self.to_json and not self.date:
    ...
else:
   ...
```

Shurke commented on rss_reader/components/news.py:
```diff
@@ -0,0 +1,87 @@
+    """This class represents a feed item"""
+
+    def __init__(self, feed_title, item, source, logger):
+        """This class constructor initializes the required variables for the news class"""
```
Глобальное замечание.
В строке документации к классу не хватает пояснения, что означает каждый аргумент.
В принципе, тоже самое касается и функций/методов.

Shurke commented on rss_reader/components/parser.py:
```diff
@@ -0,0 +1,42 @@
+        return self.parser.parse_args(argv)
+
+
+def validate_date_arg(input_value):
```
Пропущена строка документации.

Shurke commented on rss_reader/rss_reader/rss_reader.py:
```diff
@@ -0,0 +1,68 @@
+                logger.info(' Successfully completed')
+            else:
+                logger.error(' Specified URL does not contain RSS. Please check the specified URL and try again')
+    else:
+        logger.error(' Source URL not specified. Please check your input and try again')
```
Подобного плана проверки стоит проводить в самом начале работы ридера: валидность введенного адреса, ввели ли этот адрес и т.д.

spectraise replied:
Они же вроде как и проводятся в самом начале. Сначала проверяется передан ли аргумент --verbose, потом проверяется правильность значения аргумента --limit (если таковой передан), затем проверяется передан ли аргумент --date или source и только потом ридер начинает что-то делать.

Shurke replied:
Тогда почему этот `else` в самом низу?

Shurke reviewed (commented):
И не хватает тестов на новый функционал + стоит пересмотреть еще раз старые.

spectraise commented:
> > В --to-html и --to-pdf должен передаваться путь по которому он должен быть сохранен. У вас ничего не передается.
> 
> Тогда стоит сообщить пользователю о том, что он забыл путь к файлу.

Это и так сообщается. 
`rss_reader: error: argument --to-pdf: expected one argument
`

Или вы предлагаете заменить само сообщение?

spectraise committed:
Fixed imports and updated rss_reader.py

1. Imports are now sorted alphabetically
2. Added function get_data_from_url for rss_reader.py

spectraise committed:
Updated 'parser' module

1. --limit argument validation moved from 'rss_reader.py' to 'parser' module
2. Changed the way the --date argument is validated

spectraise committed:
Fixing problems

1. Added missing docstrings
2. Reworked module 'cache'
3. Reworked module 'converter'

spectraise committed:
Fixed docstring for news_limit parameter

spectraise committed:
Trying to improve tests for arguments

spectraise committed:
Updated 'test_arguments.py'

Imports are now sorted alphabetically

spectraise committed:
Fixed tests for errors

spectraise committed:
Trying to fix the path to 'data/example.xml' in 'test_arguments.py'

spectraise committed:
Updated 'test_arguments.py'

Imports are now sorted alphabetically

spectraise committed:
Added tests for 'cache' module

spectraise committed:
Added cached_news.json for 'cache' module tests

spectraise committed:
Fixed missing docstrings and naming

spectraise committed:
Updated README.md

spectraise committed:
Fixed missing docstring in 'parser' module

spectraise committed:
Fixed imports in 'rss_reader.py'

Imports are now sorted alphabetically

spectraise committed:
Updated README.md

Shurke commented:
Из Task clarification (IV):

> You can choose yourself the way in which the news will be displayed, but the final text result should contain pictures and links, if they exist in the original article and if the format permits to store this type of data.

![Screenshot from 2021-05-25 22-01-58](https://user-images.githubusercontent.com/44495358/119553984-d9b90680-bda4-11eb-8d59-ba6fbde7da39.png)
![Screenshot from 2021-05-25 22-05-42](https://user-images.githubusercontent.com/44495358/119554488-619f1080-bda5-11eb-8342-89d72c448e46.png)

spectraise committed:
Added handling of media:content elements

1. Added method '__parse_media_content' to 'news' module
2. Changed the way the content type is determined
3. Added method '__link_to_html' to 'converter' module
4. Updated template.html
5. JSON structure changed

spectraise committed:
Updated README.md

spectraise committed:
Fixed tests for 'cache' module

Shurke commented on rss_reader/rss_reader/rss_reader.py:
```diff
@@ -0,0 +1,88 @@
+    cache = Cache(logger)
+    if args.date:
+        feeds_list = cache.get_news_from_cache(args.date, args.source, args.limit, args.json)
+        if args.to_pdf is not None:
```
А почему бы не написать `if args.to_pdf:`?

Shurke commented on rss_reader/rss_reader/rss_reader.py:
```diff
@@ -0,0 +1,88 @@
+        for feed in feeds_list:
+            print(feed)
+    elif args.source:
+        soup = get_data_from_url(logger, args.source)
```
Имя переменной `soup` не несет никакого логического смысла.

Shurke commented on rss_reader/components/news.py:
```diff
@@ -0,0 +1,130 @@
+            try:
+                content_type = content['type']
+            except KeyError:
+                content_type = 'unknown'
```
Для этого у словаря есть метод `get()`
```python
content_type = content.get('type')
```

Shurke commented on rss_reader/components/news.py:
```diff
@@ -0,0 +1,130 @@
+            str: Content type
+        """
+        response = requests.head(content['url'])
+        try:
+            content_type = response.headers['content-type']
+        except KeyError:
+            content_type = None
```
Для этого у словаря есть метод get().

Shurke commented on rss_reader/components/news.py:
```diff
@@ -0,0 +1,130 @@
+            return parse(date)
+        elif self.item.published:
+            date = self.item.published.text
+            return parse(date)
```
Точно ли тут нужно 2 `return`?
```python
if self.item.pubDate:
    date = self.item.pubDate.text
elif self.item.published:
    date = self.item.published.text
return parse(date)
```

Или pubDate и published могут быть одновренно?

Shurke commented on rss_reader/components/feed.py:
```diff
@@ -0,0 +1,55 @@
+        self.cache = cache
+        if not self.news_list:
+            self.news_items = news_items
+            self.__create_feed()
```
Вот этот кусок выглядит странно. В 31 строке делается присваивание, а потом проверка.
Стоит сначала проверить `news_list`, а потом уже присваивать значения для `self.news_list` и `self.news_items`.

Shurke commented on rss_reader/tests/test_arguments.py:
```diff
@@ -0,0 +1,99 @@
+        """Tests that if --json option is specified utility should convert the news into JSON format"""
+        with patch.object(self.example_feed, 'to_json', True):
+            print(str(self.example_feed))
+        try:
+            json.loads(mock_stdout.getvalue())
+        except json.JSONDecodeError:
+            self.fail('JSONDecodeError raised by json.loads')
```
В тестах не должно быть `try/except`.

Shurke commented on rss_reader/tests/test_cache.py:
```diff
@@ -0,0 +1,73 @@
+class TestRssReader(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        cls.data_folder = os.path.abspath(os.path.join(os.path.dirname(__file__), 'data')) + os.path.sep
+        cls.cache_folder = os.path.abspath(os.path.join(os.path.dirname(__file__), 'cache')) + os.path.sep
+        with open(cls.data_folder + 'example.xml', 'r') as file:
+            cls.soup = BeautifulSoup(file.read(), 'lxml-xml')
+        cls.example_feed_title = cls.soup.find('title').text
+        cls.example_items = cls.soup.find_all('item')
+        cls.example_feed = Feed('https://www.yahoo.com/news', None, False, logging, cls.example_feed_title,
+                                Cache(logging, cls.cache_folder), news_items=cls.example_items)
+        cls.example_news_list = cls.example_feed.news_list
+        shutil.rmtree(cls.cache_folder)
```
Как я вижу, то этот код такой же, как и в rss_reader/tests/test_arguments.py.
Стоит подумать о том, чтобы создать какой-то базовый класс, где реализовать `setUpClass`.
И от этого базового класса наследовать тестовые классы в rss_reader/tests/test_cache.py и rss_reader/tests/test_arguments.py.

spectraise committed:
KeyError exception handling replaced with .get () dictionary method

spectraise committed:
Now self.news_list is assigned only after checking news_list

spectraise committed:
Updated logging functionality

1. Added command-line argument '--colorize' which now only affects log messages
2. Changed log messages format
3. Changed tests for the new format of log messages

spectraise committed:
colorlog module added to requirements

spectraise committed:
[Version 0.5] Added command-line argument '--colorize'

1. Added command-line argument '--colorize'
2. Added method 'colorize_string' to module 'news'
3. Changed method '__str__' in module 'feed'
4. Changed method '__str__' in module 'news'

spectraise committed:
Fixed naming of variables and filepath to font file

spectraise committed:
Updated tests

1. Tests no longer go online
2. Added tests for the module 'converter'

spectraise committed:
Fixed test_to_html

spectraise committed:
Added tests for --date argument validator

spectraise committed:
Updated tests

1. Removed try/except blocks
2. Reworked tests for the module 'converter'
3. Added function mock_response to testing.py
4. Pytest is now used for testing instead of nosetests

spectraise committed:
Fixed requirements

spectraise committed:
Reworked test_to_html and test_to_html_with_limit in test_converter.py
